# Анализ ассортимента детской литературы магазина Читай-Город

Итак, мы отдебажили код и собрали по частям полный (на тот момент) ассортимент детской литературы с сайта магазина Читай-Город.
Теперь объединим все файлы в один датафрейм.

In [1]:
import pandas as pd

Так как искомые файлы лежат в другой папке, найдём их с помощью библиотеки os.

In [2]:
import os

In [12]:
# Посмотрим, что лежит в папке предыдущего дз.

pp = list(os.walk('C:\\Users\\Klim\\Desktop\\Анализ Данных\\Python\\Klim_notebooks\\Homework\\Parsing_project'))
pp

[('C:\\Users\\Klim\\Desktop\\Анализ Данных\\Python\\Klim_notebooks\\Homework\\Parsing_project',
  ['.ipynb_checkpoints'],
  ['book_error.text',
   'book_error_02.text',
   'CG_Pages.ipynb',
   'CG_Pages_With_function.ipynb',
   'chg.txt',
   'HW3_Diana.ipynb',
   'Klim_chitai-gorod.ru__knigi_dlya_detey.ipynb',
   'Myshows_parsing Анна.ipynb',
   'parsed_chitay_gorod_hudlit.csv',
   'parsed_chitay_gorod_hudlit_02.csv',
   'parsed_chitay_gorod_kids.csv',
   'parsed_chitay_gorod_kids_02.csv',
   'parsed_chitay_gorod_kids_debug.csv',
   'parsed_chitay_gorod_kids_final.csv',
   'parsed_chitay_gorod_kids_missed.csv',
   'parsed_chitay_gorod_kids_missed_02.csv',
   'parsed_chitay_gorod_kids_missed_03.csv',
   'parsed_chitay_gorod_kids_missed_04.csv',
   'Parsing practice.ipynb',
   'Parsing_Klim_Chitai-Gorod.ipynb',
   'Sidekick.ipynb']),
 ('C:\\Users\\Klim\\Desktop\\Анализ Данных\\Python\\Klim_notebooks\\Homework\\Parsing_project\\.ipynb_checkpoints',
  [],
  ['CG_Pages-checkpoint.ipynb',
  

In [19]:
# Создадим список адресов интересующих нас файлов.

kids = []
for three in range(len(pp[0][2])):
    if 'kids' in pp[0][2][three] and not 'final' in pp[0][2][three]:
        kids.append(os.path.join(pp[0][0], pp[0][2][three]))
kids

['C:\\Users\\Klim\\Desktop\\Анализ Данных\\Python\\Klim_notebooks\\Homework\\Parsing_project\\parsed_chitay_gorod_kids.csv',
 'C:\\Users\\Klim\\Desktop\\Анализ Данных\\Python\\Klim_notebooks\\Homework\\Parsing_project\\parsed_chitay_gorod_kids_02.csv',
 'C:\\Users\\Klim\\Desktop\\Анализ Данных\\Python\\Klim_notebooks\\Homework\\Parsing_project\\parsed_chitay_gorod_kids_debug.csv',
 'C:\\Users\\Klim\\Desktop\\Анализ Данных\\Python\\Klim_notebooks\\Homework\\Parsing_project\\parsed_chitay_gorod_kids_missed.csv',
 'C:\\Users\\Klim\\Desktop\\Анализ Данных\\Python\\Klim_notebooks\\Homework\\Parsing_project\\parsed_chitay_gorod_kids_missed_02.csv',
 'C:\\Users\\Klim\\Desktop\\Анализ Данных\\Python\\Klim_notebooks\\Homework\\Parsing_project\\parsed_chitay_gorod_kids_missed_03.csv',
 'C:\\Users\\Klim\\Desktop\\Анализ Данных\\Python\\Klim_notebooks\\Homework\\Parsing_project\\parsed_chitay_gorod_kids_missed_04.csv']

In [22]:
# И создадим датафрейм, объединяющий все эти файлы, используя функцию pd.read_csv на список kids через map.

df = pd.concat(map(pd.read_csv, kids)).reset_index(drop=True)

C:\Users\Klim\AppData\Local\Temp\ipykernel_11308\3983233578.py:1: DtypeWarning: Columns (9) have mixed types. Specify dtype option on import or set low_memory=False.
  df = pd.concat(map(pd.read_csv, kids)).reset_index(drop=True)


Несмотря на предупреждение, датафрейм создан.

In [30]:
# Сразу удалим задвоения.

df.drop_duplicates(inplace=True)

In [38]:
# Переиндексуем.

df = df.reset_index(drop=True)

In [39]:
# Выведем основные данные.

df.shape
df.info()

<class 'pandas.core.frame.DataFrame'>
RangeIndex: 43324 entries, 0 to 43323
Data columns (total 31 columns):
 #   Column                       Non-Null Count  Dtype  
---  ------                       --------------  -----  
 0   ID товара                    43324 non-null  int64  
 1   Название                     43324 non-null  object 
 2   Автор                        43324 non-null  object 
 3   Цикл                         43324 non-null  object 
 4   Том                          43324 non-null  object 
 5   Категория                    43324 non-null  object 
 6   Подкатегории                 42684 non-null  object 
 7   Издательство                 43324 non-null  object 
 8   Серия                        43324 non-null  object 
 9   Год издания                  43324 non-null  object 
 10  Возрастные ограничения       43324 non-null  object 
 11  Переводчик                   43324 non-null  object 
 12  Художник                     43324 non-null  object 
 13  Цена            

In [40]:
# Посмотрим, как это выглядит.

df

ID товара                                          Название  \
0        2938632  Мир вокруг тебя. Первая энциклопедия для малышей   
1        2938633           Космос. Первая энциклопедия для малышей   
2        2946110              История одного волшебства (выпуск 1)   
3        2946460                    Крольчонок Миа, или Ключи зимы   
4        2950851       Тачки в пути. Большая книга раскрасок и игр   
...          ...                                               ...   
43319    2296390                                  Вежливый трактор   
43320    2294180                                Сказки для девочек   
43321    2294153                  Р Первый бал Посмотри и раскрась   
43322    2291674                                        Снегурочка   
43323    2287976                        Дневничок для супердевочек   

                       Автор Цикл Том                             Категория  \
0      Феданова Ю., Скиба Т.    -   -  Детская научно-популярная литература   
1      Феданова Ю., Скиба Т.    -   -  Детская научно-популярная литература   
2               <не указано>    -   -     Детская художественная литература   
3                  Медоус Д.    -   -     Детская художественная литература   
4                          -    -   -     Детская художественная литература   
...                      ...  ...  ..                                   ...   
43319             Бутенко К.    -   -     Детская художественная литература   
43320       Рашина Т. (ред.)    -   -     Детская художественная литература   
43321            Мигунова Н.    -   -     Детская художественная литература   
43322     Федотова М. (худ.)    -   -     Детская художественная литература   
43323           Дмитриева В.    -   -            Детское творчество и досуг   

                                        Подкатегории    Издательство  \
0           Первые энциклопедии для малышей (0-6 л.)          Владис   
1           Первые энциклопедии для малышей (0-6 л.)               -   
2                 Проза для детей. Повести, рассказы           Эксмо   
3                                             Сказки           Эксмо   
4                   Книги-игрушки/Книги с наклейками           Эксмо   
...                                              ...             ...   
43319                  Книги-игрушки/Книги - игрушки  Мозаика-Синтез   
43320                                         Сказки      Проф-пресс   
43321                                      Раскраски      Проф-пресс   
43322                  Книги-игрушки/Книги - игрушки         Карапуз   
43323  Все для детских игр/Дневники. Альбомы. Анкеты             АСТ   

                                          Серия Год издания  ...  \
0               Первая энциклопедия для малышей        2022  ...   
1               Первая энциклопедия для малышей        2022  ...   
2                                             -        2023  ...   
3      Детск. Уютные истории для зимнего вечера        2023  ...   
4      Disney. PIXAR. Игры, раскраски, наклейки        2022  ...   
...                                         ...         ...  ...   
43319                         Книжка-погремушка        2011  ...   
43320              Читаем детям. Читаем малышам        2011  ...   
43321                                 Раскраски        2011  ...   
43322                          Книжки на брюшке        2012  ...   
43323                                         -        2011  ...   

      Кол-во страниц        Тип обложки             Формат  Вес  Тираж  \
0                 64  Твердая глянцевая  24.7 x 17.5 x 0.8  226   5000   
1                 64  Твердая глянцевая  24.7 x 17.5 x 0.8  226   5000   
2                144                  -                  -    -      -   
3                192                  -                  -    -      -   
4                 64                  -                  -    -      -   
...              ...                ...                ...  ...    ...   
43319            

Итак, у нас на руках датафрейм из 43324 строк и 31 столбца.
Большинство признаков - объекты, но мы это исправим.

In [33]:
# Сохранимся в текущую рабочую директорию.

df.to_csv('parsed_chitay_gorod_kids_final.csv', index=False)

Поработаем с непредусмотренными пропусками.

## Пропуски

In [42]:
df.isnull().sum()

ID товара                        0
Название                         0
Автор                            0
Цикл                             0
Том                              0
Категория                        0
Подкатегории                   640
Издательство                     0
Серия                            0
Год издания                      0
Возрастные ограничения           0
Переводчик                       0
Художник                         0
Цена                            48
Цена со скидкой                 48
Рейтинг                          0
Кол-во оценок                    0
ISBN                             0
Аннотация                       76
Отрывок                          0
Обложка                          0
Кол-во страниц                   0
Тип обложки                      0
Формат                           0
Вес                              0
Тираж                            0
Наличие на сайте                 0
Поступит в продажу               0
Наличие в магазинах 

**Цена**

In [92]:
print('{}%'.format(round(len(df[df['Цена'].isnull() == True]) / len(df) * 100, 2)))

0.11%


Записей с пропущенной ценой всего 0,11%. Но посмотрим на эти записи подробнее.

In [75]:
for j in range(len(list(df.columns))):
    print(list(df.columns)[j], df[df['Цена'].isnull() == True][list(df.columns)[j]].unique())

ID товара [2783604 2783656 2783657 2783658 2783659 2783660 2783661 2775692 2775662
 2765883 2749145 2739083 2721175 2721177 2719265 2717980 2706852 2706659
 2706688 2706795 2706796 2706799 2706803 2706594 2706629 2702327 2702005
 2698224 2684798 2684786 2684562 2684537 2684532 2683873 2683824 2683822
 2683654 2681725 2677503 2677467 2677466 2666740 2664856 2642692 2642628
 2637258 2635978 2621030]
Название ['Холодное торжество. Выпуск 2'
 'Гравити Фолз. Раскраска № 1 (Диппер и Мэйбл)'
 'Гравити Фолз. Раскраска № 2 (Мэйбл и дядя Стэн)'
 'Гравити Фолз. Раскраска (с наклейками) № 1 (Диппер и Зус)'
 'Гравити Фолз. Раскраска (с наклейками) № 2 (Мэйбл и Венди)'
 'Холодное сердце. Сила волшебства. Книга для чтения'
 'Холодное сердце. Ветер перемен. Книга для чтения'
 'Если бы доисторические чудовища жили в наши дни. Охотники глубин'
 'Сесилия Гатэ и тайна Мальтийских часов'
 'Тролли 2. Наклей и раскрась (Цветан)' 'Дракоша Тоша - почтальон'
 'Большая книга приключений Медведя и Гусика' 'Дракош

Про эти книги можно сказать следующее:
- они не часть какого-либо цикла;
- их нет (и не предвидится) на сайте и в магазиах;
- среди них нет новинок (вышли не позднее 2020 года);
- об их форме (Формат и Вес) утеряны все сведения.

Выведем среди них отдельно книги с оценкой от 4 и выше.

In [82]:
for j in range(len(list(df.columns))):
    print(list(df.columns)[j], df[(df['Цена'].isnull() == True) &
                                  (df['Рейтинг'] >= 4)][list(df.columns)[j]].unique())

ID товара [2783656 2783657 2783658 2681725 2677467 2642692]
Название ['Гравити Фолз. Раскраска № 1 (Диппер и Мэйбл)'
 'Гравити Фолз. Раскраска № 2 (Мэйбл и дядя Стэн)'
 'Гравити Фолз. Раскраска (с наклейками) № 1 (Диппер и Зус)'
 'Старик Хоттабыч' 'Новогодние наклейки'
 'Дневник слабака. Эту книгу сделай сам']
Автор ['-' 'Лагин Л.' 'Джефф Кинни']
Цикл ['-']
Том ['-']
Категория ['Детская художественная литература' 'Детское творчество и досуг']
Подкатегории ['Раскраски' 'Сказки' 'Книги-игрушки/Книги с наклейками'
 'Все для детских игр/Дневники. Альбомы. Анкеты']
Издательство ['Эксмодетство' 'АСТ']
Серия ['Disney. Гравити Фолз. Раскраски' 'Самые настоящие мальчишки'
 '250 наклеек' 'Дневник слабака']
Год издания [2020 2018]
Возрастные ограничения ['-' '12+']
Переводчик ['-']
Художник ['-']
Цена [nan]
Цена со скидкой [nan]
Рейтинг [5.  4.5 4. ]
Кол-во оценок [1 2 4 5]
ISBN ['978-5-04-111109-0' '978-5-04-111113-7' '978-5-04-111111-3'
 '978-5-271-48223-6' '978-5-17-108810-1' '978-5-17-103043-

Вроде ничего невосполнимого.
Теперь проверим, есть ли в ассортименте другие издания про Простоквашино.

In [103]:
df[df['Название'].str.contains('Простоквашино', case=False)]

ID товара                                           Название  \
1528     2932498           Зима в Простоквашино. Новогодние истории   
1978     2920248                           Истории из Простоквашино   
5554     2880200             Старые и новые истории о Простоквашино   
6602     2870476              Зима в Простоквашино и другие истории   
10929    2825457    Дядя Федор, пес и кот. Истории из Простоквашино   
15492    2774249      Дядя Федор, пес и кот в деревне Простоквашино   
16682    2763948                Маленькие сказочки из Простоквашино   
17151    2759934                           Истории из Простоквашино   
18264    2742233  Возвращение в Простоквашино (1 кнопка с 3 песе...   
20820    2706590         Все-все-все лучшие истории о Простоквашино   
22512    2681524  Дядя Федор, пес и кот и другие истории о Прост...   
22526    2680948            Веселые истории в деревне Простоквашино   
22879    2674611  Каникулы в Простоквашино. Рисуй по точкам и ра...   
22918    2669874         Трое из Простоквашино. Книжка с закладками   
22982    2669732  Простоквашино. Кто живет в деревне. Книжка на ...   
23388    2664852  С Новым годом! Новый год с Чебурашкой. Новый г...   
23734    2660592  Домашние животные в Простоквашино. Придумай св...   
28454    2573229  Дядя Федор, пес и кот, или Кое-что о Простоква...   
28551    2569911                           Истории из Простоквашино   
29381    2540231                           Истории из Простоквашино   
29746    2520433        Маленькие сказочные истории о Простоквашино   
32673    2841705              Дядя Федор, пес и кот в Простоквашино   
33577    2787716  Творческие задачки с наклейками. Союзмультфиль...   
33717    2783403            Головоломки с наклейками. Простоквашино   
34425    2759238                           Каникулы в Простоквашино   
35821    2695027                               Зима в Простоквашино   
35824    2695028                           Каникулы в Простоквашино   
35834    2695025                              Весна в Простоквашино   
35978    2691803       Домашние животные в простоквашино. Раскраска   
36246    2681644                 Маленькие истории из Простоквашино   
36314    2677831                               Зима в Простоквашино   
36426    2674948         В Простоквашино. Союзмультфильм. Раскраска   
36496    2670129  Союзмультфильм. Домашние животные в Простокваш...   
36654    2669862  Простоквашино. Домашние животные. Книжка-пищал...   
36897    2665034                                  Все Простоквашино   
37111    2660521           Трое из Простоквашино. Рисуем пальчиками   
37698    2642859                            Сказки из Простоквашино   
39148    2581158                    Веселые истории в Простоквашино   
39783    2546607  Альбом наклеек. Домашние животные в Простоквашино   
40351    2501670                           Истории из Простоквашино   
40681    2478691       Каникулы в Простоквашино. Наклей и раскрась!   
40688    2478646  Домашние животные в Простоквашино. Наклей и ра...   
40695    2478608            Трое из Простоквашино. Книжка-панорамка   
40771    2471797             Старые и новые истории о Простоквашино   
41099    2446109   Большая книга о Простоквашино. Сказочные повести   
43017    2683822              Зима в Простоквашино и другие истории   

                                         Автор Цикл Том  \
1528   Успенский Э.Н., Остер Г.Б., Маршак С.Я.    -   -   
1978                            Успенский Э.Н.    -   -   
5554                                Э.Успенсий    -   -   
6602                              Э. Успенский    -   -   
10929                             Э. Успенский    -   -   
15492                             Э. Успенский    -   -   
16682                             Э. Успенский    -   -   
17151                             Э. Успенский    -   -   
18264                       Хомякова К. (ред.)    -   -   
20820                             Э. Успенский    -   -   
22512              У

Изданий предостаточно. С чистой совестью удаляем пропуски по цене.

In [279]:
df.drop(df[df['Цена'].isnull() == True].index, axis=0, inplace=True)

Теперь цену и цену со скидкой можно преобразовать в int.

In [151]:
df['Цена'] = df['Цена'].astype('int64')
df['Цена со скидкой'] = df['Цена со скидкой'].astype('int64')

Заодно преобразуем ID товара в строку, так как это не число.

In [153]:
df['ID товара'] = df['ID товара'].astype('str')

In [154]:
df.info()

<class 'pandas.core.frame.DataFrame'>
Int64Index: 43276 entries, 0 to 43323
Data columns (total 31 columns):
 #   Column                       Non-Null Count  Dtype  
---  ------                       --------------  -----  
 0   ID товара                    43276 non-null  object 
 1   Название                     43276 non-null  object 
 2   Автор                        43276 non-null  object 
 3   Цикл                         43276 non-null  object 
 4   Том                          43276 non-null  object 
 5   Категория                    43276 non-null  object 
 6   Подкатегории                 42636 non-null  object 
 7   Издательство                 43276 non-null  object 
 8   Серия                        43276 non-null  object 
 9   Год издания                  43276 non-null  object 
 10  Возрастные ограничения       43276 non-null  object 
 11  Переводчик                   43276 non-null  object 
 12  Художник                     43276 non-null  object 
 13  Цена            

**Аннотация**

Аннотация не является обязательным признаком для книги, поэтому просто заменим пропуски на '-'.

In [163]:
df['Аннотация'] = df['Аннотация'].fillna('-')

**Подкатегории**

Найдём категории пропущенных подкатегорий.

In [183]:
cats = list(df[df['Подкатегории'].isnull() == True]['Категория'].unique())
cats

['Детское творчество и досуг',
 'Детская религиозная литература',
 'Детская художественная литература']

Определим, сколько всего подкатегорий в каждой из выделенных категорий.

In [196]:
df[df['Категория'].isin(cats)].groupby('Категория')['Подкатегории'].nunique()

Категория
Детская религиозная литература        0
Детская художественная литература    17
Детское творчество и досуг           11
Name: Подкатегории, dtype: int64

Оказывается, детская религиозная литература вообще не разбита на подкатегории,
а для двух других категорий только некоторые книги не выделены в особую подкатегорию. 

Посчитаем эти книги.

In [225]:
cats2 = dict(df[(df['Категория'].isin(cats)) & 
                (df['Подкатегории'].isnull() == True)].groupby('Категория')['Название'].count())
cats2

{'Детская религиозная литература': 530,
 'Детская художественная литература': 95,
 'Детское творчество и досуг': 15}

Определим процент таких книг в каждой выделенной категории.

In [226]:
print('Не выделено в подкатегории, % книг: \n')
for cat in cats2:
    print(cat + ': ' + '{}%'.format(round(cats2[cat] / len(df[df['Категория'] == cat]) * 100, 2)))

Не выделено в подкатегории, % книг: 

Детская религиозная литература: 100.0%
Детская художественная литература: 0.29%
Детское творчество и досуг: 0.32%


И, так как ничего страшного тут нет, просто присвоим значение признака 'Категория' признаку 'Подкатегории' каждой такой книги.

Но сначала сохранимся в другой файл.

In [233]:
df.to_csv('parsed_chitay_gorod_kids_final_02.csv', index=False)

In [274]:
df.loc[df[df['Подкатегории'].isnull() == True].index, 
       'Подкатегории'] = df.loc[df[df['Подкатегории'].isnull() == True].index, 'Категория']

Что-то произошло. Давайте посмотрим, что именно.

In [275]:
df['Подкатегории'].isnull().sum()

0

In [276]:
df[df['Подкатегории'] == df['Категория']].shape[0]

640

Пустых подкатегорий больше нет и приняли значение признака 'Категория' ровно 640 штук.

In [277]:
df.isnull().sum()

ID товара                      0
Название                       0
Автор                          0
Цикл                           0
Том                            0
Категория                      0
Подкатегории                   0
Издательство                   0
Серия                          0
Год издания                    0
Возрастные ограничения         0
Переводчик                     0
Художник                       0
Цена                           0
Цена со скидкой                0
Рейтинг                        0
Кол-во оценок                  0
ISBN                           0
Аннотация                      0
Отрывок                        0
Обложка                        0
Кол-во страниц                 0
Тип обложки                    0
Формат                         0
Вес                            0
Тираж                          0
Наличие на сайте               0
Поступит в продажу             0
Наличие в магазинах            0
Наличие в местных магазинах    0
Карта     

Пропусков больше нет, кроме прочерков, которые были предусмотрены на этапе сбора информации.

In [281]:
df.info()

<class 'pandas.core.frame.DataFrame'>
Int64Index: 43276 entries, 0 to 43323
Data columns (total 31 columns):
 #   Column                       Non-Null Count  Dtype  
---  ------                       --------------  -----  
 0   ID товара                    43276 non-null  object 
 1   Название                     43276 non-null  object 
 2   Автор                        43276 non-null  object 
 3   Цикл                         43276 non-null  object 
 4   Том                          43276 non-null  object 
 5   Категория                    43276 non-null  object 
 6   Подкатегории                 43276 non-null  object 
 7   Издательство                 43276 non-null  object 
 8   Серия                        43276 non-null  object 
 9   Год издания                  43276 non-null  object 
 10  Возрастные ограничения       43276 non-null  object 
 11  Переводчик                   43276 non-null  object 
 12  Художник                     43276 non-null  object 
 13  Цена            

In [273]:
df.loc[df[df['Том'].str.contains(',')].index, 'Название']

12322                             Драконья сага. Мятежники
15215                                       Хроники Нарнии
15216                                       Хроники Нарнии
21070                          Все о муми-троллях. Книга 1
22568                          Макабр (комплект из 3 книг)
24141                                   Пеппи Длинныйчулок
24366                       Вампир Полумракс. Звездный пес
25149                      Маленькие женщины. Хорошие жены
27162    Все о Волшебнике Изумрудного Города. Сказочные...
27602                                       Хроники Нарнии
28048         Хроники Нарнии. Последняя битва. Три повести
28376                                       Хроники Нарнии
29033           Волшебник Изумрудного города и все-все-все
29302                   Пеппи Длинныйчулок. Повесть-сказка
30261      Все о девочке с Земли. Приключения продолжаются
30473    Гарри Поттер. Полное собрание (комплект из 7 к...
31918    Хроники Нарнии: начало истории. Четыре повести.